File: components/api/api-modules-javascript/src/main/resources/META-INF/dirigible/modules/src/io/zip.ts
/*
 * Copyright (c) 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 * SPDX-License-Identifier: EPL-2.0
 */
/**
 * API Zip
 */
import { Bytes } from "sdk/io/bytes";
import { InputStream, OutputStream } from "sdk/io/streams";

const ZipFacade = Java.type("org.eclipse.dirigible.components.api.io.ZipFacade");

export class Zip {

    public static zip(sourcePath: string, zipTargetPath: string): void {
        ZipFacade.exportZip(sourcePath, zipTargetPath);
    }

    public static unzip(zipPath: string, targetPath: string): void {
        ZipFacade.importZip(zipPath, targetPath);
    }

    public static createZipInputStream(inputStream: InputStream): ZipInputStream {
        const native = ZipFacade.createZipInputStream(inputStream.native);
        return new ZipInputStream(native);
    }

    public static createZipOutputStream(outputStream: OutputStream): ZipOutputStream {
        const native = ZipFacade.createZipOutputStream(outputStream.native);
        return new ZipOutputStream(native);
    }
}

export class ZipInputStream {

    private readonly native: any;

    constructor(native: any) {
        this.native = native;
    }

    public getNextEntry(): ZipEntry {
        const native = this.native.getNextEntry();
        return new ZipEntry(native);
    }

    public read(): any[] {
        const native = ZipFacade.readNative(this.native);
        return Bytes.toJavaScriptBytes(native);
    }

    public readNative(): any[] {
        return ZipFacade.readNative(this.native);
    }

    public readText(): string {
        return ZipFacade.readText(this.native);
    }

    public close(): void {
        this.native.close();
    }

}

export class ZipOutputStream {

    private readonly native: any;

    constructor(native: any) {
        this.native = native;
    }

    public createZipEntry(name: string): ZipEntry {
        const nativeNext = ZipFacade.createZipEntry(name);
        const zipEntry = new ZipEntry(nativeNext);
        this.native.putNextEntry(nativeNext);
        return zipEntry;
    }

    public write(data: any[]): void {
        const native = Bytes.toJavaBytes(data);
        ZipFacade.writeNative(this.native, native);
    }

    public writeNative(data: any[]): void {
        ZipFacade.writeNative(this.native, data);
    }

    public writeText(text: string): void {
        ZipFacade.writeText(this.native, text);
    }

    public closeEntry(): void {
        this.native.closeEntry();
    }

    public close(): void {
        this.native.finish();
        this.native.flush();
        this.native.close();
    }
}

/**
 * ZipEntry object
 */
export class ZipEntry {

    private readonly native: any;

    constructor(native: any) {
        this.native = native;
    }

    public getName(): string {
        return this.native.getName();
    }

    public getSize(): number {
        return this.native.getSize();
    }

    public getCompressedSize(): number {
        return this.native.getCompressedSize();
    }

    public getTime(): number {
        return this.native.getTime();
    }

    public getCrc(): number {
        return this.native.getCrc();
    }

    public getComment(): string {
        return this.native.getComment();
    }

    public isDirectory(): boolean {
        return this.native.isDirectory();
    }

    public isValid(): boolean {
        return this.native !== undefined && this.native !== null;
    }
}

// @ts-ignore
if (typeof module !== 'undefined') {
	// @ts-ignore
	module.exports = Zip;
}

File: components/api/api-modules-java/src/main/java/org/eclipse/dirigible/sdk/io/Zip.java
/*
 * Copyright (c) 2010-2026 Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: Eclipse Dirigible contributors SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.dirigible.sdk.io;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;
import org.eclipse.dirigible.components.api.io.ZipFacade;

/**
 * ZIP archive creation and extraction. Two styles are supported:
 *
 * <ul>
 * <li>The folder-level shortcuts {@link #importZip(String, String)} (extract everything under a
 * directory) and {@link #exportZip(String, String)} (zip a folder onto disk).</li>
 * <li>The streaming form via {@link ZipInputStream} / {@link ZipOutputStream} when you want
 * entry-by-entry control — useful for producing a download on the fly, scanning archives without
 * writing intermediate files, or filtering entries during extraction.</li>
 * </ul>
 *
 * For end-to-end "zip this folder, send it" workflows the shortcuts are enough; reach for the
 * streaming overloads only when you need entry-level control.
 */
public final class Zip {

    private Zip() {}

    public static void importZip(String zipPath, String targetPath) {
        ZipFacade.importZip(zipPath, targetPath);
    }

    public static void exportZip(String folderPath, String zipPath) {
        ZipFacade.exportZip(folderPath, zipPath);
    }

    public static ZipInputStream createZipInputStream(InputStream in) throws IOException {
        return ZipFacade.createZipInputStream(in);
    }

    public static ZipOutputStream createZipOutputStream(OutputStream out) throws IOException {
        return ZipFacade.createZipOutputStream(out);
    }

    public static ZipEntry createZipEntry(String name) throws IOException {
        return ZipFacade.createZipEntry(name);
    }

    public static void write(ZipOutputStream output, byte[] bytes) throws IOException {
        ZipFacade.write(output, bytes);
    }

    public static void write(ZipOutputStream output, String data) throws IOException {
        ZipFacade.write(output, data);
    }

    public static void writeNative(ZipOutputStream output, byte[] data) throws IOException {
        ZipFacade.writeNative(output, data);
    }

    public static void writeText(ZipOutputStream output, String text) throws IOException {
        ZipFacade.writeText(output, text);
    }

    public static String read(ZipInputStream input) throws IOException {
        return ZipFacade.read(input);
    }

    public static byte[] readNative(ZipInputStream input) throws IOException {
        return ZipFacade.readNative(input);
    }

    public static String readText(ZipInputStream input) throws IOException {
        return ZipFacade.readText(input);
    }
}
